File: javaScript/medium/bstTraversal.js
// O(n) time | O(n) space

//                      10
//                    5     15
//                  2   5       22
//                1

function inOrderTraverse(tree, array) {
  // [1, 2, 5, 5, 10, 15, 22]
  if (tree !== null) {
    inOrderTraverse(tree.left, array);
    array.push(tree.value);
    inOrderTraverse(tree.right, array);
  }
  return array;
}

function preOrderTraverse(tree, array) {
  // [10, 5, 2, 1, 5, 15, 22]
  if (tree !== null) {
    array.push(tree.value);
    preOrderTraverse(tree.left, array);
    preOrderTraverse(tree.right, array);
  }
  return array;
}

function postOrderTraverse(tree, array) {
  // [1, 2, 5, 5, 22, 15, 10]
  if (tree !== null) {
    postOrderTraverse(tree.left, array);
    postOrderTraverse(tree.right, array);
    array.push(tree.value);
  }
  return array;
}

File: python/Medium/bst-traversal.py

# BST TRAVERSAL

# ALL THREE METHODS: O(N) time and space
def inOrderTraverse(tree, array):
    # Write your code here.
	if tree:
		inOrderTraverse(tree.left, array)
		array.append(tree.value)
		inOrderTraverse(tree.right, array)
	return array
    
def preOrderTraverse(tree, array):
    # Write your code here.
    if tree:
		array.append(tree.value)
		preOrderTraverse(tree.left, array)
		preOrderTraverse(tree.right, array)
	return array

def postOrderTraverse(tree, array):
    # Write your code here.
    if tree:
		postOrderTraverse(tree.left, array)
		postOrderTraverse(tree.right, array)
		array.append(tree.value)
	return array